src/server_manager/web_app/app.ts

```diff
@@ -200,136 +209,218 @@ export class App {
       this.appRoot.openGetConnectedDialog(this.getS3InviteUrl(event.detail.accessKey, true));
     });
 
+    appRoot.addEventListener('ShowServerRequested', (event: PolymerEvent) => {
+      const displayServerId = event.detail.displayServerId;
+      const serverEntry = this.serverDisplayTable[displayServerId];
+      if (!serverEntry && !!this.serverBeingCreated) {
+        this.showServerCreationProgress();
+      } else if (!!serverEntry) {
+        this.showServerIfHealthy(serverEntry.server, serverEntry.displayServer);
+      } else {
+        // We should never reach this if the server display table is synced.
+        console.error(`Could not find server entry for display server ID ${displayServerId}`);
+      }
+    });
+
     onUpdateDownloaded(this.displayAppUpdateNotification.bind(this));
   }
 
   start(): void {
-    // Load manual servers from storage.
-    this.manualServerRepository.listServers().then((manualServers) => {
-      // Show any manual servers if they exist.
-      if (manualServers.length > 0) {
-        this.showManualServerIfHealthy(manualServers[0]);
-        return;
-      }
+    // Load display servers and associate them with managed and manual servers.
+    this.displayServerRepository.listServers().then((displayServers) => {
+      this.appRoot.serverList = displayServers;
+
+      this.manualServerRepository.listServers().then((manualServers) => {
+        this.syncServersToDisplay(manualServers).then(() => {
+          this.maybeShowLastDisplayedServer();
+        });
+      });
 
-      // User has no manual servers - check if they are logged into DigitalOcean.
       const accessToken = this.digitalOceanTokenManager.getStoredToken();
       if (accessToken) {
-        this.enterDigitalOceanMode(accessToken);
-        return;
+        this.enterDigitalOceanMode(accessToken).then((managedServers) => {
+          this.syncServersToDisplay(managedServers).then(() => {
+            this.maybeShowLastDisplayedServer();
+          });
+        });
       }
+    });
+    this.showIntro();
+  }
 
-      // User has no manual servers or DigitalOcean token.
-      this.showIntro();
+  private async syncServersToDisplay(servers: server.Server[]) {
+    for (const server of servers) {
+      await this.syncServerToDisplay(server);
+    }
+  }
+
+  // Syncs the locally persisted server metadata for `server`. Creates a DisplayServer for `server`
+  // if one is not found in storage.
+  private async syncServerToDisplay(server: server.Server): Promise<DisplayServer> {
```

Comment:
I think this function tries to do too much. Seems like we need functions to:
- query a server and return a `DisplayServer`
- update the display table
- add new items to the display table

Breaking those functions out will make things more readable overall and help reduce some of the duplicate code currently in this function, viz. multiple calls to `syncDisplayServersToUi``.